File: src/main/java/org/hildan/minecraft/mining/optimizer/patterns/AbstractDiggingPattern.java
package org.hildan.minecraft.mining.optimizer.patterns;

import org.hildan.minecraft.mining.optimizer.chunks.Block;
import org.hildan.minecraft.mining.optimizer.chunks.Explorer;
import org.hildan.minecraft.mining.optimizer.chunks.Sample;
import org.hildan.minecraft.mining.optimizer.chunks.Wrapping;

import java.util.ArrayList;
import java.util.Collection;

/**
 * This abstract class regroups common behavior for all patterns.
 */
public abstract class AbstractDiggingPattern implements DiggingPattern {

    @Override
    public void digInto(Sample sample) {
        Collection<Access> accesses = new ArrayList<>();
        for (int x = 0; x < sample.getWidth(); x += getWidth()) {
            for (int y = 0; y < sample.getHeight(); y += getHeight()) {
                accesses.addAll(getAccesses(x, y));
                for (int z = 0; z < sample.getLength(); z += getLength()) {
                    digInto(sample, x, y, z);
                }
            }
        }
        Explorer.explore(sample, accesses);
        digVisibleOres(sample);
    }

    private static void digVisibleOres(Sample sample) {
        sample.getBlocksMatching(b -> b.isOre() && b.isVisible()).forEach(b -> digBlockAndAdjacentOres(sample, b));
    }

    private static void digBlockAndAdjacentOres(Sample sample, Block block) {
        sample.digBlock(block.getX(), block.getY(), block.getZ());
        Collection<Block> adjacentBlocks = sample.getAdjacentBlocks(block, Wrapping.CUT);
        for (Block adjacentBlock : adjacentBlocks) {
            if (adjacentBlock.isOre() && adjacentBlock.isVisible()) {
                digBlockAndAdjacentOres(sample, adjacentBlock);
            }
        }
    }

    /**
     * Digs this pattern into the given sample, starting from the given origin, and going in the increasing direction of
     * each coordinate. This method takes care of stopping at the edge of the given sample.
     */
    protected abstract void digInto(Sample sample, int originX, int originY, int originZ);

    @Override
    public String toString() {
        Sample sample = new Sample(getWidth(), getHeight(), getLength());
        digInto(sample);
        return sample.toString();
    }
}

File: src/main/kotlin/org/hildan/minecraft/mining/optimizer/patterns/AbstractDiggingPattern.kt
package org.hildan.minecraft.mining.optimizer.patterns

import org.hildan.minecraft.mining.optimizer.blocks.Block
import org.hildan.minecraft.mining.optimizer.blocks.BlockType
import org.hildan.minecraft.mining.optimizer.blocks.Explorer
import org.hildan.minecraft.mining.optimizer.blocks.Sample
import org.hildan.minecraft.mining.optimizer.blocks.Wrapping

/**
 * This abstract class regroups common behavior for all patterns.
 */
abstract class AbstractDiggingPattern : DiggingPattern {

    override fun digInto(sample: Sample) {
        val accesses = mutableListOf<Access>()
        for (x in 0 until sample.width step width) {
            for (y in 0 until sample.height step height) {
                accesses.addAll(getAccesses(x, y))
                for (z in 0 until sample.length step length) {
                    digInto(sample, x, y, z)
                }
            }
        }
        Explorer.explore(sample, accesses)
        digVisibleOres(sample)
    }

    private fun digVisibleOres(sample: Sample) {
        sample.getBlocksMatching { it.isOre && it.isVisible }.forEach { digBlockAndAdjacentOres(sample, it) }
    }

    private fun digBlockAndAdjacentOres(sample: Sample, block: Block) {
        sample.digBlock(block.x, block.y, block.z)
        val adjacentBlocks = sample.getAdjacentBlocks(block, Wrapping.CUT)
        for (adjacentBlock in adjacentBlocks) {
            if (adjacentBlock.isOre && adjacentBlock.isVisible) {
                digBlockAndAdjacentOres(sample, adjacentBlock)
            }
        }
    }

    /**
     * Digs this pattern into the given sample, starting from the given origin, and going in the increasing direction of
     * each coordinate. This method must take care of stopping at the edge of the given sample.
     */
    protected abstract fun digInto(sample: Sample, originX: Int, originY: Int, originZ: Int)

    override fun toString(): String {
        val sample = Sample(width, height, length, BlockType.STONE)
        digInto(sample)
        return sample.toString()
    }
}
